Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp

```diff
@@ -145,117 +83,54 @@ void IOChannel::read(size_t offset, size_t size, WTF::WorkQueueBase& queue, Func
         return;
     }
 
-    if (!RunLoop::isMain()) {
-        readSyncInThread(offset, size, queue, WTFMove(completionHandler));
-        return;
-    }
-
-    size_t bufferSize = std::min(size, gDefaultReadBufferSize);
-    uint8_t* bufferData = static_cast<uint8_t*>(fastMalloc(bufferSize));
-    GRefPtr<GBytes> buffer = adoptGRef(g_bytes_new_with_free_func(bufferData, bufferSize, fastFree, bufferData));
-    ReadAsyncData* asyncData = new ReadAsyncData { this, buffer.get(), queue, size, WTFMove(completionHandler), { } };
-
-    // FIXME: implement offset.
-    g_input_stream_read_async(m_inputStream.get(), bufferData, bufferSize, RunLoopSourcePriority::DiskCacheRead, nullptr,
-        reinterpret_cast<GAsyncReadyCallback>(inputStreamReadReadyCallback), asyncData);
-}
-
-void IOChannel::readSyncInThread(size_t offset, size_t size, WTF::WorkQueueBase& queue, Function<void(Data&, int error)>&& completionHandler)
-{
-    ASSERT(!RunLoop::isMain());
-
-    Thread::create("IOChannel::readSync", [this, protectedThis = Ref { *this }, size, queue = Ref { queue }, completionHandler = WTFMove(completionHandler)] () mutable {
-        size_t bufferSize = std::min(size, gDefaultReadBufferSize);
-        uint8_t* bufferData = static_cast<uint8_t*>(fastMalloc(bufferSize));
-        GRefPtr<GBytes> readBuffer = adoptGRef(g_bytes_new_with_free_func(bufferData, bufferSize, fastFree, bufferData));
-        Data data;
-        size_t pendingBytesToRead = size;
-        size_t bytesToRead = bufferSize;
-        do {
-            // FIXME: implement offset.
-            gssize bytesRead = g_input_stream_read(m_inputStream.get(), const_cast<void*>(g_bytes_get_data(readBuffer.get(), nullptr)), bytesToRead, nullptr, nullptr);
-            if (bytesRead == -1) {
-                queue->dispatch([protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)] {
-                    Data data;
-                    completionHandler(data, -1);
-                });
-                return;
+    Thread::create("IOChannel::read", [this, protectedThis = WTFMove(protectedThis), offset, size, queue = Ref { queue }, completionHandler = WTFMove(completionHandler)]() mutable {
+        GRefPtr<GFileInfo> info = adoptGRef(g_file_input_stream_query_info(G_FILE_INPUT_STREAM(m_inputStream.get()), G_FILE_ATTRIBUTE_STANDARD_SIZE, nullptr, nullptr));
+        if (!info) {
+            auto fileSize = g_file_info_get_size(info.get());
+            if (fileSize && fileSize <= static_cast<goffset>(std::numeric_limits<size_t>::max())) {
```

Comment:
But goffset is signed, so `static_cast<goffset>(std::numeric_limits<size_t>::max())` is going to be negative. Won't the `fileSize <=` check always fail?